consensus/istanbul/uptime/uptime.go

```diff
@@ -0,0 +1,242 @@
+		numerator := big.NewInt(0).Mul(big.NewInt(int64(entry.ScoreTally)), params.Fixidity1)
+		uptimes = append(uptimes, big.NewInt(0).Div(numerator, big.NewInt(int64(denominator))))
+	}
+
+	if len(uptimes) < valSetSize {
+		err := fmt.Errorf("%d accumulated uptimes found, cannot update validator scores", len(uptimes))
+		logger.Error(err.Error())
+		return nil, err
+	}
+
+	return uptimes, nil
+}
+
+// ProcessBlock will monitor uptimeness information on the give block and add it to current epoch tally
+func (um *Monitor) ProcessBlock(block *types.Block) error {
+	// The epoch's first block's aggregated parent signatures is for the previous epoch's valset.
+	// We can ignore updating the tally for that block.
+	if istanbul.IsFirstBlockOfEpoch(block.NumberU64(), um.epochSize) {
+		return nil
+	}
+
+	// Get the bitmap from the previous block
+	extra, err := types.ExtractIstanbulExtra(block.Header())
+	if err != nil {
+		um.logger.Error("Unable to extract istanbul extra", "func", "ProcessBlock", "blocknum", block.NumberU64())
+		return errors.New("could not extract block header extra")
+	}
+	signedValidatorsBitmap := extra.ParentAggregatedSeal.Bitmap
+
+	// Get the uptime scores
+	epochNum := istanbul.GetEpochNumber(block.NumberU64(), um.epochSize)
+	uptime := um.store.ReadAccumulatedEpochUptime(epochNum)
+
+	// We only update the uptime for blocks which are greater than the last block we saw.
+	// This ensures that we do not count the same block twice for any reason.
+	if uptime == nil || uptime.LatestBlock < block.NumberU64() {
+		uptime = updateUptime(uptime, block.NumberU64(), signedValidatorsBitmap, um.lookbackWindow, epochNum, um.epochSize)
+		um.store.WriteAccumulatedEpochUptime(epochNum, uptime)
+	} else {
+		log.Trace("WritingBlockWithState with block number less than a block we previously wrote", "latestUptimeBlock", uptime.LatestBlock, "blockNumber", block.NumberU64())
+	}
+
+	return nil
+}
+
+// updateUptime receives the currently accumulated uptime for all validators in an epoch and proceeds to update it
+// based on which validators signed on the provided block by inspecting the block's parent bitmap
+//
+// It expects a blockNumber with the signatures bitmap from the previous block (parent seal)
+func updateUptime(uptime *Uptime, blockNumber uint64, bitmap *big.Int, window uint64, epochNum uint64, epochSize uint64) *Uptime {
+	if uptime == nil {
+		uptime = new(Uptime)
+		// The number of validators is upper bounded by 3/2 of the number of 1s in the bitmap
+		// We multiply by 2 just to be extra cautious of off-by-one errors.
+		validatorsSizeUpperBound := uint64(math.Ceil(float64(bitCount(bitmap)) * 2))
+		uptime.Entries = make([]UptimeEntry, validatorsSizeUpperBound)
+	}
+
+	firstBlockToMonitor, lastBlockToMonitor := GetUptimeMonitoringWindow(epochNum, epochSize, window)
+
+	// lookbackBlockNum := (blockNumber -1) - window
+
+	// signedBlockWindowLastBlockNum is just the previous block
+	signedBlockWindowLastBlockNum := blockNumber - 1
```

Comment:
while we're renaming things, would this be clearer as `lookbackWindowLastBlockNum` and `lookbackWindowLastBlockNum`